Venafi issuer setup: verify access token

jahrlin opened the pull request:
<!--

Thanks for opening a pull request! Here are some tips to get everything merged smoothly:

1. Read our contributor guidelines: https://cert-manager.io/docs/contributing/

2. Make sure your commits are signed off: https://cert-manager.io/docs/contributing/sign-off/

3. If the PR is unfinished, raise it as a draft or prefix the title with "WIP:" so it's clear to everyone.

4. Be sure to allow edits from maintainers so it's easier for us to help: https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork

-->

Fixes #5154 

This is the first PR that fixes the immediate problem: a Venafi Issuer does not verify its access token or API key.

This PR changes the Venafi Issuer to remotely verify its access token (TPP) or API key (Cloud) with a remote server during setup. An invalid (expired, malformed) access token or API key will result in the Issuer setting `Ready=False`.

A follow-up PR will add functionality to periodically periodically perform this check.

Follow-up issue here: https://github.com/cert-manager/cert-manager/issues/5201

### Kind

bug

<!--

Pick a kind which best describes your PR from the following list:

	<cleanup | bug | feature | documentation | design | flake>

If you're unsure which is best or if you're not sure what we mean by "kind",
just ignore this section and a maintainer will fill it in for you!
-->

### Release Note

<!--

Should we mention this PR in release notes? If so, replace "NONE" with a line of text explaining what changed!

For more details, see: https://git.k8s.io/community/contributors/guide/release-notes.md

-->

```release-note
Fixed a bug where the Venafi Issuer would not verify its access token (TPP) or API key (Cloud) before becoming ready. Venafi Issuers will now remotely verify the access token or API key.. (#5154, @jahrlin)
```

jahrlin commented:
/test all

jahrlin commented on pkg/issuer/venafi/client/venaficlient.go:
```diff
@@ -124,6 +138,14 @@ func configForIssuer(iss cmapi.GenericIssuer, secretsLister corelisters.SecretLi
 				Password:    password,
 				AccessToken: accessToken,
 			},
+			// this is needed for local development when tunneling to the TPP server
```
I'd like to keep this in here for the future

irbekrm commented:
/test all

jahrlin commented:
/test all

jahrlin commented:
/test pull-cert-manager-e2e-v1-23

jahrlin commented:
/test pull-cert-manager-issuers-venafi-tpp

jahrlin commented:
/test pull-cert-manager-issuers-venafi-tpp

jahrlin commented:
/test pull-cert-manager-issuers-venafi-tpp

jahrlin commented:
/test all

jahrlin commented:
/test pull-cert-manager-e2e-v1-23

jahrlin commented:
/test pull-cert-manager-e2e-v1-23

irbekrm commented:
/test pull-cert-manager-e2e-v1-24

irbekrm reviewed (changes requested):
Thank you for working on this @jahrlin !

I have added a couple comments. I'm a little uneasy with adding TPP-specific code at the top to be run for both VaaS and TPP issuers. I think ideally we should have some generic functionality that check auth for Venafi issuer and runs the TPP-specific bits in TPP-specific blocks so we don't have to expose TPP-specific fields at top level. I also think that ideally these calls should be instrumented, not sure if achieving both goals will be easy.
Let me know what you think.

Additionally I think the issuer's Ready condition should definitely be set to False if this check fails- do I understand it correctly that currently the code is not doing that? (I am not even sure if we want to return an error in the reconciler- I imagine that in most cases this is an error that will require changes to the `Secret` resource which will then cause a resync. But perhaps there are some edge cases where this failure can be fixed by changing something in TPP?)

A very minor nit: we don't really have a policy for this, but it subjectively I think that it might be good to squash the commits here. I personally usually utilize rewriting commits and force pushing quite heavily to ensure that each commit is one logical change

Question: have we tested this? Given that the e2e tests actually don't test with TPP token at all it could be nice to add to the PR description a screenshot or something that shows this functionality in action

Also it looks like this is currently broken for VaaS https://prow.build-infra.jetstack.net/view/gs/jetstack-logs/pr-logs/pull/cert-manager_cert-manager/5185/pull-cert-manager-e2e-issuers-venafi-cloud/1535290360698245120

jahrlin commented:
/test pull-cert-manager-e2e-issuers-venafi-cloud
/test pull-cert-manager-issuers-venafi-tpp

jahrlin commented:
/test pull-cert-manager-e2e-issuers-venafi-cloud
/test pull-cert-manager-issuers-venafi-tpp

jahrlin commented:
/test pull-cert-manager-issuers-venafi-tpp

jahrlin commented:
/test pull-cert-manager-e2e-issuers-venafi-cloud
/test pull-cert-manager-issuers-venafi-tpp

jahrlin commented:
/test pull-cert-manager-e2e-v1-24
/test pull-cert-manager-issuers-venafi-tpp

jahrlin commented:
> I have added a couple comments. I'm a little uneasy with adding TPP-specific code at the top to be run for both VaaS and TPP issuers. I think ideally we should have some generic functionality that check auth for Venafi issuer and runs the TPP-specific bits in TPP-specific blocks so we don't have to expose TPP-specific fields at top level. I also think that ideally these calls should be instrumented, not sure if achieving both goals will be easy. Let me know what you think.

✅  I have added a generic function to verify the credentials for both Cloud and TPP issuers, including tests 

❌  Using the instrumentedclient is not doable without a "larger" rewrite, which is something we can do but in a separate PR. The issue is we are depending on vcerts' endpoint.Connector-interface in a few places and I can't see a straightforward path to extending the instrumentedclient with this functionality without a "larger" rewrite.

> Additionally I think the issuer's Ready condition should definitely be set to False if this check fails- do I understand it correctly that currently the code is not doing that? (I am not even sure if we want to return an error in the reconciler- I imagine that in most cases this is an error that will require changes to the `Secret` resource which will then cause a resync. But perhaps there are some edge cases where this failure can be fixed by changing something in TPP?)

✅ The issuers Ready condition will be set to False if this check fails.

> A very minor nit: we don't really have a policy for this, but it subjectively I think that it might be good to squash the commits here. I personally usually utilize rewriting commits and force pushing quite heavily to ensure that each commit is one logical change

✅ I prefer squashing as well, ideally letting GitHub do it for me - I can hold this and squash myself when approved.

> Question: have we tested this? Given that the e2e tests actually don't test with TPP token at all it could be nice to add to the PR description a screenshot or something that shows this functionality in action

✅ Yes, I've tested this locally using an access token both manually and while running e2e tests.

> Also it looks like this is currently broken for VaaS https://prow.build-infra.jetstack.net/view/gs/jetstack-logs/pr-logs/pull/cert-manager_cert-manager/5185/pull-cert-manager-e2e-issuers-venafi-cloud/1535290360698245120

✅ Should be happy now

jahrlin commented on pkg/issuer/venafi/setup.go:
```diff
@@ -46,13 +45,19 @@ func (v *Venafi) Setup(ctx context.Context) (err error) {
 	}
 
-	res, err := client.VerifyAccessToken(&endpoint.Authentication{})
+	err = client.VerifyAccessToken()
```
See my other comment on this part

irbekrm commented:
Thanks for adding the checks for VaaS and username/password auth and the extra tests @jahrlin !

I've read through the code and this all makes sense to me.

I think we might want to update the release note and the PR description to include the fact that TPP username/password and VaaS access key will also be checked.

>  Using the instrumentedclient is not doable without a "larger" rewrite, which is something we can do but in a separate PR. The issue is we are depending on vcerts' endpoint.Connector-interface in a few places and I can't see a straightforward path to extending the instrumentedclient with this functionality without a "larger" rewrite.

Yes that makes sense, I think we probably will have to do that though. Probably worth creating an issue for this.

/lgtm
/approve

jahrlin commented:
> I think we might want to update the release note and the PR description to include the fact that TPP username/password and VaaS access key will also be checked.

✅  Done!

> > Using the instrumentedclient is not doable without a "larger" rewrite, which is something we can do but in a separate PR. The issue is we are depending on vcerts' endpoint.Connector-interface in a few places and I can't see a straightforward path to extending the instrumentedclient with this functionality without a "larger" rewrite.
> 
> Yes that makes sense, I think we probably will have to do that though. Probably worth creating an issue for this.

✅ Done: https://github.com/cert-manager/cert-manager/issues/5210

jahrlin commented:
Fixed in https://github.com/cert-manager/cert-manager/pull/5212
